client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/ShuffleResourceTracker.java

```diff
@@ -0,0 +1,210 @@
+  private final ExecutorService executorService;
+  private final LifecycleManager lifecycleManager;
+  // JobID -> ShuffleResourceListener
+  private final Map<JobID, JobShuffleResourceListener> shuffleResourceListeners =
+      new ConcurrentHashMap<>();
+  private static final int MAX_RETRY_TIMES = 3;
+
+  public ShuffleResourceTracker(
+      ExecutorService executorService, LifecycleManager lifecycleManager) {
+    this.executorService = executorService;
+    this.lifecycleManager = lifecycleManager;
+    lifecycleManager.registerWorkerStatusListener(this);
+  }
+
+  public void registerJob(JobShuffleContext jobShuffleContext) {
+    shuffleResourceListeners.put(
+        jobShuffleContext.getJobId(),
+        new JobShuffleResourceListener(jobShuffleContext, executorService));
+  }
+
+  public void addPartitionResource(
+      JobID jobId, int shuffleId, int partitionId, ResultPartitionID partitionID) {
+    JobShuffleResourceListener shuffleResourceListener = shuffleResourceListeners.get(jobId);
+    shuffleResourceListener.addPartitionResource(shuffleId, partitionId, partitionID);
+  }
+
+  public void removePartitionResource(JobID jobID, int shuffleId, int partitionId) {
+    JobShuffleResourceListener shuffleResourceListener = shuffleResourceListeners.get(jobID);
+    if (shuffleResourceListener != null) {
+      shuffleResourceListener.removePartitionResource(shuffleId, partitionId);
+    }
+  }
+
+  public JobShuffleResourceListener getJobResourceListener(JobID jobID) {
+    return shuffleResourceListeners.get(jobID);
+  }
+
+  public void unRegisterJob(JobID jobID) {
+    shuffleResourceListeners.remove(jobID);
+  }
+
+  @Override
+  public void notifyChangedWorkersStatus(WorkersStatus workersStatus) {
+    try {
+      List<WorkerInfo> unknownWorkers = workersStatus.unknownWorkers;
+      if (unknownWorkers != null && !unknownWorkers.isEmpty()) {
+        // untrack by job
+        for (Map.Entry<JobID, JobShuffleResourceListener> entry :
+            shuffleResourceListeners.entrySet()) {
+          Set<ResultPartitionID> partitionIds = new HashSet<>();
+          JobShuffleResourceListener shuffleResourceListener = entry.getValue();
+          for (Map.Entry<Integer, Map<Integer, ResultPartitionID>> mapEntry :
+              shuffleResourceListener.getResultPartitionMap().entrySet()) {
+            int shuffleId = mapEntry.getKey();
+            if (!mapEntry.getValue().isEmpty()) {
+              for (WorkerInfo unknownWorker : unknownWorkers) {
+                Map<WorkerInfo, ShufflePartitionLocationInfo> shuffleAllocateInfo =
+                    lifecycleManager.workerSnapshots(shuffleId);
+                // shuffleResourceListener may release when the shuffle is ended
+                if (shuffleAllocateInfo != null) {
+                  ShufflePartitionLocationInfo shufflePartitionLocationInfo =
+                      shuffleAllocateInfo.remove(unknownWorker);
```

Comment:
> 

In fact we can't remove here, since the later loop will find shufflePartitionLocationInfo is null, which is not correct. Maybe just use get here.